src/category_theory/conj.lean

```diff
@@ -0,0 +1,142 @@
+/-
+Copyright (c) 2019 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+
+import category_theory.isomorphism data.equiv.basic category_theory.endomorphism algebra.group_power
+
+/-!
+# Conjugate morphisms by isomorphisms
+
+An isomorphism `α : X ≅ Y` defines
+- a monoid isomorphism `conj : End X ≃* End Y` by `α.conj f = α.inv ≫ f ≫ α.hom`;
+- a group isomorphism `conj_Aut : Aut X ≃* Aut Y` by `α.conj_Aut f = α.symm ≪≫ f ≪≫ α`.
+
+For completeness, we also define `hom_congr : (X ≅ X₁) → (Y ≅ Y₁) → (X ⟶ Y) ≃ (X₁ ⟶ Y₁)`, cf. `equiv.arrow_congr`.
+-/
+
+universes v u
+
+namespace category_theory
+
+namespace iso
+
+variables {C : Type u} [𝒞 : category.{v+1} C]
+include 𝒞
+
+/- If `X` is isomorphic to `X₁` and `Y` is isomorphic to `Y₁`, then
+there is a natural bijection between `X ⟶ Y` and `X₁ ⟶ Y₁`. See also `equiv.arrow_congr`. -/
+def hom_congr {X Y X₁ Y₁ : C} (α : X ≅ X₁) (β : Y ≅ Y₁) :
+  (X ⟶ Y) ≃ (X₁ ⟶ Y₁) :=
+{ to_fun := λ f, α.inv ≫ f ≫ β.hom,
+  inv_fun := λ f, α.hom ≫ f ≫ β.inv,
+  left_inv := λ f, show α.hom ≫ (α.inv ≫ f ≫ β.hom) ≫ β.inv = f,
+    by rw [category.assoc, category.assoc, β.hom_inv_id, α.hom_inv_id_assoc, category.comp_id],
+  right_inv := λ f, show α.inv ≫ (α.hom ≫ f ≫ β.inv) ≫ β.hom = f,
+    by rw [category.assoc, category.assoc, β.inv_hom_id, α.inv_hom_id_assoc, category.comp_id] }
+
+lemma hom_congr_apply {X Y X₁ Y₁ : C} (α : X ≅ X₁) (β : Y ≅ Y₁) (f : X ⟶ Y) :
+  α.hom_congr β f = α.inv ≫ f ≫ β.hom :=
+rfl
+
+lemma hom_congr_comp {X Y Z X₁ Y₁ Z₁ : C} (α : X ≅ X₁) (β : Y ≅ Y₁) (γ : Z ≅ Z₁)
+  (f : X ⟶ Y) (g : Y ⟶ Z) :
+  α.hom_congr γ (f ≫ g) = (α.hom_congr β f) ≫ (hom_congr β γ g) :=
+by simp only [hom_congr_apply, category.assoc, β.hom_inv_id_assoc]
+
+@[simp] lemma hom_congr_refl {X Y : C} (f : X ⟶ Y) :
+  (iso.refl X).hom_congr (iso.refl Y) f = f :=
+by simp only [hom_congr_apply, iso.refl, category.comp_id, category.id_comp]
+
+@[simp] lemma hom_congr_trans {X₁ Y₁ X₂ Y₂ X₃ Y₃ : C}
+  (α₁ : X₁ ≅ X₂) (β₁ : Y₁ ≅ Y₂) (α₂ : X₂ ≅ X₃) (β₂ : Y₂ ≅ Y₃) (f : X₁ ⟶ Y₁) :
+  (α₁ ≪≫ α₂).hom_congr (β₁ ≪≫ β₂) f = (α₁.hom_congr β₁).trans (α₂.hom_congr β₂) f :=
+by simp only [hom_congr_apply, equiv.trans_apply, iso.trans, category.assoc]
+
+@[simp] lemma hom_congr_symm {X₁ Y₁ X₂ Y₂ : C} (α : X₁ ≅ X₂) (β : Y₁ ≅ Y₂) :
+  (α.hom_congr β).symm = α.symm.hom_congr β.symm :=
+rfl
+
+variables {X Y : C} (α : X ≅ Y)
+
+/-- An isomorphism between two objects defines a monoid isomorphism between their
+monoid of endomorphisms. -/
+def conj : End X ≃* End Y :=
+{ map_mul' := λ f g, hom_congr_comp α α α g f,
+  .. hom_congr α α }
+
+@[simp] lemma conj_apply (f : End X) : α.conj f = α.inv ≫ f ≫ α.hom := rfl
```

Comment:
This is a good lemma, but I'm not sure if it's a good simp-lemma. For example, I'm not sure if I would want this to trigger either before or after `conj_pow` below. Maybe priorities can help? I'm not sure what exactly would be desirable here.